forcedotcom/source-deploy-retrieve
# Do not attempt string replacements on binary files

### What does this PR do?
If you attempt to do string replacements on a binary file (e.g. an image) it will become corrupt due to the `.toString()` conversion. This change prevents string replacements from running on binary files. 

[@W-16560716@](https://gus.my.salesforce.com/apex/ADM_WorkLocator?bugorworknumber=W-16560716)

## Testing
- Create a salesforce project and create a scratch org
- Either manually or from the Setup UI, add a `staticresource` image
  - If you created it in the UI, `retrieve` it to your local project
- Add this config to your `sfdx-project.json`
```json
"replacements": [
    {
      "glob": "force-app/main/default/**/**",
      "stringToReplace": "replaceMe",
      "replaceWithEnv": "FOO"
    }
  ]
```
- Run `export FOO="bar"` or it will complain that the env is not set
- Before this PR, run `sf project deploy start --source-dir force-app/main/default/staticresources`
- You will notice that the image shows as `changed`
- Go to the `staticresource` in the browser and you will notice that the size is much larger and it will also fail to render if you click `View file`
- **Note: Do you `retrieve` again or your local image will become corrupt**
- Now pull this branch, `yarn link` it in `plugin-deploy-retrieve`, and `sf plugins link .` PDR
- Re-run `sf project deploy start --source-dir force-app/main/default/staticresources`
- You will see that it shows as `changed` again, but this is because it is replacing the corrupt version
- Check out the image in the UI, it should load.
- Run `sf project deploy start --source-dir force-app/main/default/staticresources` on final time and it should show as `unchanged`

## Notes
Here are some screenshots showing the file size differences. There values are used in the NUTs to confirm the file did not change.
OK:
<img width="482" alt="Screenshot 2024-08-22 at 2 18 38 PM" src="https://github.com/user-attachments/assets/02f7a8d7-1b20-45ba-b6cd-0038c8233147">

Corrupt:
<img width="483" alt="Screenshot 2024-08-22 at 2 20 31 PM" src="https://github.com/user-attachments/assets/0a9bb79e-8d25-4e29-8084-ee9e1398202c">

Proof from test:
<img width="1064" alt="Screenshot 2024-08-22 at 2 17 58 PM" src="https://github.com/user-attachments/assets/029f91b9-5bae-40f7-bb78-7c9717297b55">
